src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.ServerSentEvents
+{
+    internal static class Helpers
+    {
+        public static unsafe void WriteAsUtf8String(this IBufferWriter<byte> bufferWriter, ReadOnlySpan<char> value)
+        {
+            if (value.IsEmpty)
+            {
+                return;
+            }
+
+            int maxByteCount = Encoding.UTF8.GetMaxByteCount(value.Length);
+            Span<byte> buffer = bufferWriter.GetSpan(maxByteCount);
+            int bytesWritten;
+#if NET
+            bytesWritten = Encoding.UTF8.GetBytes(value, buffer);
+#else
+            fixed (char* chars = value)
+            fixed (byte* bytes = buffer)
+            {
+                bytesWritten = Encoding.UTF8.GetBytes(chars, value.Length, bytes, maxByteCount);
+            }
+#endif
+            bufferWriter.Advance(bytesWritten);
+        }
+
+        public static void ValidateParameterDoesNotContainLineBreaks(string? input, string paramName)
+        {
+            if (input?.Contains('\n') is true)
+            {
+                Throw(paramName);
+                static void Throw(string parameterName) => throw new ArgumentException(SR.ArgumentException_MustNotContainLineBreaks, parameterName);
+            }
+        }
+
+#if !NET
+        public static bool Contains(this string text, char character) => text.IndexOf(character) >= 0;
+
+        public static ValueTask WriteAsync(this Stream stream, ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            if (MemoryMarshal.TryGetArray(buffer, out ArraySegment<byte> segment))
+            {
+                return new ValueTask(stream.WriteAsync(segment.Array, segment.Offset, segment.Count, cancellationToken));
+            }
+            else
+            {
+                return WriteAsyncUsingPooledBuffer(stream, buffer, cancellationToken);
+
+                static async ValueTask WriteAsyncUsingPooledBuffer(Stream stream, ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+                {
+                    byte[] sharedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length);
+                    buffer.Span.CopyTo(sharedBuffer);
+                    try
+                    {
+                        await stream.WriteAsync(sharedBuffer, 0, buffer.Length, cancellationToken).ConfigureAwait(false);
+                    }
+                    finally
```

Comment:
> The array pool should handle dropped arrays in case of an exception.

How is that?

Reply:
In case of an exception the array won't be returned to the pool, so ultimately it will be GCed.
That's no problem for the array pool, so the `try-finally` isn't needed.

Reply:
The exception here isn't the common case, so one can avoid additional complexity brought by try-finally that is not needed.

The pattern with try-finally for the array pool was used at the beginning of AP, but recently renting / returning is just done w/o try-finally. See e.g. https://github.com/dotnet/runtime/issues/71249#issuecomment-1165486832 and lots of other comments for this.

Reply:
I don't think inserting a try/finally is so terrible from a code complexity standpoint. We use that pattern pervasively and I don't think we need to remove it here.

Reply:
> We use that pattern pervasively

Really? Recently I see it the other way around.

But to shorten this discussion to whatever you think is right here.

Reply:
I don't at all agree with that approach. Should we be starving the buffer pool just because of a transient network outage?

Reply:
async operations aren't necessarily done referencing the buffer when they throw (it's possible the native half can write after the managed half throws a timeout exception, for example), so they're not supposed to return it except in successful flows.

Reply:
> it's possible the native half can write after the managed half throws a timeout exception, for example

Can you give an example in the repo where that might happen?

Reply:
Not off the top of my head.  I remember it being A Thing that Levi was pushing in 2919 or earlier (I see a commit I made in 2019 getting rid of a bunch of async finally-returns); but I can't find the written version of that.

Something involving patterns where you call a native equivalent of BeginRead(state, dest, onEndCallback), then Task cancellation tries to cancel the OS async operation, but that causes a race condition with the IO completion event and the async operation cancellation event.

@GrabYourPitchforks might be able to spout one off... or perhaps correct my mis-saved "fact"